Email

Nightsd01 opened the pull request:
• Adds email to the iOS SDK
     • Adds `setEmail:`, `setUnauthenticatedEmail:`, `logoutEmail:`, and some convenience methods
     • Adds a separate email observer class & subscription state model
     • Adds tests to cover email functionality
     • Adds HTTP request re-attempts when receiving status code 500~
     • Adds email to the Swift demo
     • Adds 'carrier name' as a parameter on registration using reflection to access core telephony framework if available

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/336)
<!-- Reviewable:end -->

jkasten2 commented:
1. Can we add an on_focus test? There wasn't a test for this before but would be good to add one since we this part of the code was touched. 

2. Needs a test for on_session to make sure it gets called for both email and player ids.

3. OneSignal.sendPurchases should also be sent to the email player_id if we have one.

---

Review status: 0 of 40 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[iOS_SDK/OneSignalSDK/Source/OneSignal.m, line 519 at r1](https://reviewable.io:443/reviews/onesignal/onesignal-ios-sdk/336#-L5AayKj8yifQ6MGPJ-L:-L5AayKj8yifQ6MGPJ-M:b-70ai53) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/395a5e2e10631628f0b2877aa08d5a55220fa5d8/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L519)):*
> ```objc
> 
> +(void)downloadIOSParams {
>     NSLog(@"DOWNLOADING IOS PARAMS");
> ```

Lets use our `onesignal_Log:ONE_S_LL_DEBUG` for this.

---

*[iOS_SDK/OneSignalSDK/Source/OneSignal.m, line 529 at r1](https://reviewable.io:443/reviews/onesignal/onesignal-ios-sdk/336#-L5Ab9DPpDb8kcs8OXZu:-L5Ab9DPpDb8kcs8OXZv:be4sjzc) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/395a5e2e10631628f0b2877aa08d5a55220fa5d8/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L529)):*
> ```objc
>                 delayedParameters = nil;
>             }
>             NSLog(@"REQUIRES EMAIL AUTH: %i", self.currentEmailSubscriptionState.requiresEmailAuth);
> ```

Remove logging

---

*[iOS_SDK/OneSignalSDK/Source/OSEmailSubscription.h, line 6 at r1](https://reviewable.io:443/reviews/onesignal/onesignal-ios-sdk/336#-L5AiOqsvMHbFjp2OLVJ:-L5AiOqsvMHbFjp2OLVK:b-8lolds) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/395a5e2e10631628f0b2877aa08d5a55220fa5d8/iOS_SDK/OneSignalSDK/Source/OSEmailSubscription.h#L6)):*
> ```objc
> //
> //  Created by Brad Hesse on 2/7/18.
> //  Copyright © 2018 Hiptic. All rights reserved.
> ```

Change Copyright OneSignal, matching our other files.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/onesignal/onesignal-ios-sdk/336#-:-L5BRIq4XQ1LCgJLszdR:b-hnci87)*
<!-- Sent from Reviewable.io -->

Nightsd01 commented:
Added tests, changed SendPurchases to also send these events to the email Player ID as well (if it exists)

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/onesignal/onesignal-ios-sdk/336#-:-L5FhUH3RT_TwKXWzMoc:b-hhtzqb)*
<!-- Sent from Reviewable.io -->

jkasten2 commented:
Reviewed 31 of 40 files at r1, 1 of 2 files at r2, 9 of 9 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/onesignal/onesignal-ios-sdk/336)*
<!-- Sent from Reviewable.io -->
